File: helpers/pyband/pyband/transaction.py
import base64
import json
from .client import Client
from typing import List, Optional
from .wallet import PublicKey
from .constant import MAX_MEMO_CHARACTERS
from .message import Msg


class Transaction:
    def __init__(self):
        self.msgs: List[Msg] = []
        self.account_num: Optional[int] = None
        self.sequence: Optional[int] = None
        self.chain_id: Optional[str] = None
        self.fee: int = 0
        self.gas: int = 200000
        self.memo: str = ""

    def with_messages(self, *msgs: Msg) -> "Transaction":
        self.msgs.extend(msgs)
        return self

    def with_auto(self, client: Client) -> "Transaction":
        if len(self.msgs) == 0:
            raise ValueError("messsage is empty, please use with_messages at least 1 message")

        addr = self.msgs[0].get_sender()
        account = client.get_account(addr)
        self.account_num = account.account_number
        self.sequence = account.sequence
        return self

    def with_account_num(self, account_num: int) -> "Transaction":
        self.account_num = account_num
        return self

    def with_sequence(self, sequence: int) -> "Transaction":
        self.sequence = sequence
        return self

    def with_chain_id(self, chain_id: str) -> "Transaction":
        self.chain_id = chain_id
        return self

    def with_fee(self, fee: int) -> "Transaction":
        self.fee = fee
        return self

    def with_gas(self, gas: int) -> "Transaction":
        self.gas = gas
        return self

    def with_memo(self, memo: str) -> "Transaction":
        if len(memo) > MAX_MEMO_CHARACTERS:
            raise ValueError("memo is too large")

        self.memo = memo
        return self

    def get_sign_data(self) -> bytes:
        if len(self.msgs) == 0:
            raise ValueError("message is empty")

        if self.account_num is None:
            raise ValueError("account_num should be defined")

        if self.sequence is None:
            raise ValueError("sequence should be defined")

        if self.chain_id is None:
            raise ValueError("chain_id should be defined")

        for msg in self.msgs:
            msg.validate()

        message_json = {
            "chain_id": self.chain_id,
            "account_number": str(self.account_num),
            "fee": {
                "gas": str(self.gas),
                "amount": [{"amount": str(self.fee), "denom": "uband"}],
            },
            "memo": self.memo,
            "sequence": str(self.sequence),
            "msgs": [x.as_json() for x in self.msgs],
        }

        message_str = json.dumps(message_json, separators=(",", ":"), sort_keys=True)
        return message_str.encode("utf-8")

    def get_tx_data(self, signature: bytes, pubkey: PublicKey) -> dict:
        return {
            "msg": [x.as_json() for x in self.msgs],
            "fee": {
                "gas": str(self.gas),
                "amount": [{"denom": "uband", "amount": str(self.fee)}],
            },
            "memo": self.memo,
            "signatures": [
                {
                    "signature": base64.b64encode(signature).decode("utf-8"),
                    "pub_key": {
                        "type": "tendermint/PubKeySecp256k1",
                        "value": base64.b64encode(bytes.fromhex(pubkey.to_hex())).decode("utf-8"),
                    },
                    "account_number": str(self.account_num),
                    "sequence": str(self.sequence),
                }
            ],
        }

File: helpers/bandchain2.js/src/transaction.ts
import { Client } from 'index'
import { Msg } from './message'
import { PublicKey } from './wallet'

export default class Transaction {
  msgs: Msg[] = []
  accountNum?: number
  sequence?: number
  chainID?: string
  fee: number = 0
  gas: number = 200000
  memo: string = ''

  withMessages(...msg: Msg[]): Transaction {
    this.msgs.push(...msg)
    return this
  }

  async withAuto(client: Client): Promise<Transaction> {
    if (this.msgs.length == 0)
      throw Error(
        'Message is empty, please use withMessages at least 1 message.',
      )

    let addr = this.msgs[0].getSender()
    let account = await client.getAccount(addr)
    if (!account) throw Error(`Account doesn't exist.`)
    this.accountNum = account.accountNumber
    this.sequence = account.sequence
    return this
  }

  withAccountNum(accountNum: number): Transaction {
    if (!Number.isInteger(accountNum)) {
      throw Error('accountNum is not an integer')
    }
    this.accountNum = accountNum
    return this
  }

  withSequence(sequence: number): Transaction {
    if (!Number.isInteger(sequence)) {
      throw Error('sequence is not an integer')
    }
    this.sequence = sequence
    return this
  }

  withChainID(chainID: string): Transaction {
    this.chainID = chainID
    return this
  }

  withFee(fee: number): Transaction {
    if (!Number.isInteger(fee)) {
      throw Error('fee is not an integer')
    }
    this.fee = fee
    return this
  }

  withGas(gas: number): Transaction {
    if (!Number.isInteger(gas)) {
      throw Error('gas is not an integer')
    }
    this.gas = gas
    return this
  }

  withMemo(memo: string): Transaction {
    this.memo = memo
    return this
  }

  getSignData(): Buffer {
    if (this.msgs.length == 0) {
      throw Error('message is empty')
    }

    if (this.accountNum == null) {
      throw Error('accountNum should be defined')
    }

    if (this.sequence == null) {
      throw Error('sequence should be defined')
    }

    if (this.chainID == null) {
      throw Error('chainID should be defined')
    }

    this.msgs.forEach((msg) => msg.validate())

    let messageJson: { [key: string]: any } = {
      chain_id: this.chainID,
      account_number: this.accountNum.toString(),
      fee: {
        amount: [
          {
            amount: this.fee.toString(),
            denom: 'uband',
          },
        ],
        gas: this.gas.toString(),
      },
      memo: this.memo,
      sequence: this.sequence.toString(),
      msgs: this.msgs.map((msg) => msg.asJson()),
    }

    const sortedKey = Object.keys(messageJson).sort()
    const result: { [key: string]: any } = {}
    sortedKey.forEach((key) => (result[key] = messageJson[key]))

    return Buffer.from(JSON.stringify(result))
  }

  getTxData(signature: Buffer, pubkey: PublicKey): Object {
    if (this.accountNum == null) {
      throw Error('accountNum should be defined')
    }

    if (this.sequence == null) {
      throw Error('sequence should be defined')
    }

    return {
      fee: {
        amount: [{ amount: this.fee.toString(), denom: 'uband' }],
        gas: this.gas.toString(),
      },
      memo: this.memo,
      msg: this.msgs.map((msg) => msg.asJson()),
      signatures: [
        {
          signature: signature.toString('base64'),
          pub_key: {
            type: 'tendermint/PubKeySecp256k1',
            value: Buffer.from(pubkey.toHex(), 'hex').toString('base64'),
          },
          account_number: this.accountNum.toString(),
          sequence: this.sequence.toString(),
        },
      ],
    }
  }
}
